python/revng/tupletree/__init__.py

```diff
@@ -370,3 +387,374 @@ def init_check_kw_only(self, *args, **kwargs):
         original_init(self, *args, **kwargs)
 
     base_class.__init__ = init_check_kw_only  # type: ignore
+
+
+def construct_type(type_info: dict, data):
```

Comment:
I added some `docstrings` and edited the code a bit. Tbh I don't know how to make this more readable without changing the underling logic. Regarding the `hasattr`, I agree is a bit annoying, but I can't get around it since we are checking for object attributes dynamically.

Reply:
@mrjackv Can you take this over?
Do we need `hasattr`?

Reply:
The current version of `construct_type` does not use `hasattr`. Should I close this?